src/tools/ceph-dencoder/ceph_dencoder.cc

```diff
@@ -75,9 +75,11 @@ class DencoderPlugin {
     }
   }
   ~DencoderPlugin() {
+#if !defined(__FreeBSD__)
```

Comment:
why?

Reply:
@tchaikov 
It looks like there are two attempts to unload the Plugins:
1. Explicit in the destructor
2. In de `exit()` sequence of things.
    Probably because it is also done in the `atexit()` table of things to do.

Without this all `ceph_dencoder` runs crash with a segmentation error.
And sometimes it also generates `__cxa_thread_call_dtors` which indicates that destructor functions are called for dynamic libraries that are already unload.

So the sollution is to not unload dl's, but to let `exit()` fix that.

Reply:
i see. i think `__cxa_thread_call_dtors` should be called by the libc for cleaning up the thread-local variables. in this case, the `vector<DencoderPlugin>` 's dtor is called as expected.

but somehow, `__cxa_thread_atexit` also tries to `dlclose()` the loaded DSOs. i took a while investigating the behavior and even came across a mail you sent to bsdmailinglist. see https://groups.google.com/g/bsdmailinglist/c/22ncTZAbDp4. and there are some more discussions on topic. see https://lists.freebsd.org/pipermail/freebsd-questions/2007-December/165687.html.

but it's still unclear what the best way to address this. your commit message explains the symptom. but it does not include the root cause analysis. could you update it with some analysis you performed, and fix the typo? i guess "due" should be "do".

Reply:
@tchaikov 
Yes, I've asked around on previous occassions. 

And this was the key answer for me:
https://groups.google.com/g/bsdmailinglist/c/22ncTZAbDp4/m/Dii_pII5AwAJ
_The FreeBSD implementation here looks racy. If one thread dlcloses an object while another thread is exiting, we can end up calling a function at an invalid memory address. It also looks as if it may be possible to unload one library, load another at the same address, and end up executing entirely the wrong code, which would have some serious security implications.

The GNU/Linux equivalent of this function locks the DSO in memory until all references to it have gone away. A call to dlclose() on GNU/Linux will not actually unload the library until all threads with destructors in that library have been unloaded. I believe that this reuses the same reference counting mechanism that allows the same library to be dlopened and dlclosed multiple times.

It would be nice if the FreeBSD version had the same behaviour, because this is almost certainly expected in code written on other platforms._

And since in the case of ceph-dencoder we are going to exit, it is the easy way out to no `dlclose()` but to let èxit()` figure it out. Other than making FreeBSD behave like Linux in this case, which is not something I feel up to ATM.

I'll put the above link and text in the commit.